### ISSUE TITLE
Document how parameter sets will hang-off hygeo objects

### ISSUE THREAD

**Issue by dblodgett-usgs** [dblodgett-usgs/hygeo#17]
Need a JSON structure that keys off catchment identifier. 

The core schema of the objects should include a model formulation type. 

Other details of the object are dictated by the formulation type.

How this relates to a catchment network is an open question. Most likely, it comes down to specifying inflow and outflow locations to the catchment. These are ostensibly contracted nodes.

There is a two-tier system here -- backbone contracted nodes and model i/o points that may or may not be coincident with contracted nodes would be represented as hydrologic locations along mainstems.

**Comment by dblodgett-usgs** [comment]
There are basically three ways that a set of catchment parameters can be associated to a catchment. 
1. associate directly to a catchment-area id
1. associate to an inflow nexus and outflow nexus
1. associate to a set of inflow nexuses and set of outflow nexuses.

The first and second are worth pursuing near term. The third is an interesting concept but might be more complexity than its worth.

For flowpaths, hydrodynamic models will either be catchment oriented or need a nexus-based coupling. 
1. associate directly to a flowpath id 
1. associate to one or more hydrologic locations that realize nexuses.

**Comment by dblodgett-usgs** [comment]
Looking at the GIUH Parameters we have so far, we have:

![Screen Shot 2020-07-28 at 3 04 58 PM](https://user-images.githubusercontent.com/1492803/88715914-e072e080-d0e3-11ea-926e-88cb925e7f2b.png)

Which is keyed off NHDPlusV2 catchment ID. 

I think what we want is a json structure like:

```json
{
  "model-id": {
    "catchment": "cat-id",
    "type": "giuh",
    "CDF": {
      "CumulativeFreq": [1, 2, 3],
      "Time": [1, 2, 3]
    },
    "minHydro": {
      "Runoff": [1, 2, 3],
      "Time": [1, 2, 3]
    },
    "hr_Hydro": {
      "Runoff": [1, 2, 3],
      "Time": [1, 2, 3]
    }
  }
}
```

or:
```json
{
  "model-id": {
    "inflow-nexus": "nex-id",
    "outflow-nexus": "nex-id",
    "type": "giuh",
    "CDF": {
      "CumulativeFreq": [1, 2, 3],
      "Time": [1, 2, 3]
    },
    "minHydro": {
      "Runoff": [1, 2, 3],
      "Time": [1, 2, 3]
    },
    "hr_Hydro": {
      "Runoff": [1, 2, 3],
      "Time": [1, 2, 3]
    }
  }
}
```

The first implies an inflow and outflow nexus, so really the second is all-encompassing.

**Comment by hellkite500** [comment]
Would something like this be what you are looking for?
https://github.com/NOAA-OWP/ngen/blob/master/data/example_realization_config.json

**Comment by dblodgett-usgs** [comment]
That's it.

**Comment by christophertubbs** [comment]
Just as a heads up: the config is going to change _just_ a bit in order in incorporate stuff like the giuh parameters within the model definition; it just makes it easier to pass as one big chunk rather than having to isolate "Ok, I have the basic model parameters and the forcing config, now to generate a giuh config and pass that in as well".

TL;DR: the giuh object is going to float up under `"tshirt"` in the config, rather than being on the same level.

**Comment by dblodgett-usgs** [comment]
## Definitions
**model domain**: A subset of the reference domain. 

- Specified as a **catchment aggregate** using an **inflow nexus** and **outflow nexus**. 
- A **model domain** uses a consistent **catchment area** **formulation** for all **contained catchments**. 
- A **model domain** is a **catchment network** with internal i/o **hydrologic locations** that link to **reference domain** i/o **hydrologic locations**. 

**reference domain**: A continental-scale hydrofabric

- A common hydrography **catchment** / **nexus** network for i/o locations and forcing / coupling interface standardization.

**catchment**: see [HY\_Features](https://docs.opengeospatial.org/is/14-111r6/14-111r6.html#_catchment)

**catchment aggregate**: [see HY\_Features](https://docs.opengeospatial.org/is/14-111r6/14-111r6.html#_catchment_aggregate)

- In this context, an aggregate of reference domain **catchments**

**inflow nexus**: [see HY\_Features](https://docs.opengeospatial.org/is/14-111r6/14-111r6.html#figure25)

**outflow nexus**: [see HY\_Features](https://docs.opengeospatial.org/is/14-111r6/14-111r6.html#figure25)

**catchment network**: [see HY\_Features](https://docs.opengeospatial.org/is/14-111r6/14-111r6.html#_catchment_network_topology)

**hydrologic location**: [see HY\_Features](https://docs.opengeospatial.org/is/14-111r6/14-111r6.html#_hydro_logic_location)


## Model Domain Parameters
For a given **model domain**, a parameter configuration file declares an **inflow nexus**, **outflow nexus**, default parameters and formulation.

For the **model domain**, mappings from model i/o **hydrologic locations** to **reference domain** i/o hydrologic locations.

For the **model domain**, a list of **catchments area** parameter objects corresponding to the **model domain** formulation must be included. The parameter objects can be empty if default parameters are to be used. A **catchment** ID is required for all **catchments** in the model domain.

**Comment by dblodgett-usgs** [comment]
After discussion today, this basic layout for parameters seemed sensible. Will work this up over in the ngen repo and propose some follow on work in Trello.
